src/providers/rateLimitedProvider.ts

```diff
@@ -0,0 +1,92 @@
+// StaticJsonRpcProvider is used in place of JsonRpcProvider to avoid redundant eth_chainId queries prior to each
+// request. This is safe to use when the back-end provider is guaranteed not to change.
+// See https://docs.ethers.io/v5/api/providers/jsonrpc-provider/#StaticJsonRpcProvider
+
+import { QueueObject, queue } from "async";
+import { ethers } from "ethers";
+import { RateLimitTask } from "./utils";
+import { getOriginFromURL } from "../utils/NetworkUtils";
+import { Logger } from "winston";
+import { Logger as umaLogger } from "@uma/logger";
+
+// This provider is a very small addition to the StaticJsonRpcProvider that ensures that no more than `maxConcurrency`
+// requests are ever in flight. It uses the async/queue library to manage this.
+export class RateLimitedProvider extends ethers.providers.StaticJsonRpcProvider {
+  // The queue object that manages the tasks.
+  private queue: QueueObject<RateLimitTask>;
+
+  // Takes the same arguments as the JsonRpcProvider, but it has an additional maxConcurrency value at the beginning
+  // of the list.
+  constructor(
+    maxConcurrency: number,
+    readonly pctRpcCallsLogged: number,
+    readonly logger: Logger = umaLogger,
+    ...cacheConstructorParams: ConstructorParameters<typeof ethers.providers.StaticJsonRpcProvider>
+  ) {
+    super(...cacheConstructorParams);
+
+    // This sets up the queue. Each task is executed by calling the superclass's send method, which fires off the
+    // request. This queue sends out requests concurrently, but stops once the concurrency limit is reached. The
+    // maxConcurrency is configured here.
+    this.queue = queue(async ({ sendArgs, resolve, reject }: RateLimitTask, callback: () => void) => {
+      await this.wrapSendWithLog(...sendArgs)
+        .then(resolve)
+        .catch(reject);
+      // we need this for the queue to know that the task is done
+      // @see: https://caolan.github.io/async/v3/global.html
+      callback();
```

Comment:
Are you sure nothing else changed with the upgrade to async? Do we have any simple tests to validate that this still works? Note: I would be okay with just a simple test to just validate our assumption that async works the way we expect. We don't have to test the full provider.

Reply:
Tests for both the queue with a callback and one without: https://github.com/across-protocol/sdk/pull/693/commits/35d2ceda2905f5d911013d799ddadab0b5163e3a